sphericart-jax/python/sphericart/jax/sph_ops.py

```diff
@@ -0,0 +1,148 @@
+    return (ShapedArray(shape, dtype), ShapedArray(dshape, dtype))
+
+
+def sph_lowering_cpu(ctx, ls, normalized, xyz):
+    """Define the compilation to XLA of the primitive.
+
+    `ctx` is a context object, the other arguments are just the regular
+    inputs.
+
+    """
+    # build shapes and dtype of the input and output
+    # arguments of the cpp functions
+    x_type = ir.RankedTensorType(xyz.type)
+    x_shape = x_type.shape
+    dtype = x_type.element_type
+    n_sample = x_shape[0]
+    ls_type = ir.RankedTensorType(ls.type)
+    ls_shape = ls_type.shape
+    l_max = ls_shape[0] - 1
+    sph_size = (l_max + 1) * (l_max + 1)
+    shape = (n_sample, sph_size)
+    dshape = (n_sample, 3, sph_size)
+
+    if dtype == ir.F32Type.get():
+        op_name = "cpu_sph_with_gradients_f32"
+    elif dtype == ir.F64Type.get():
+        op_name = "cpu_sph_with_gradients_f64"
+    else:
+        raise NotImplementedError(f"Unsupported dtype {dtype}")
+
+    return custom_call(
+        op_name,
+        # Output types
+        out_types=[
+            mlir.ir.RankedTensorType.get(shape, dtype),
+            mlir.ir.RankedTensorType.get(dshape, dtype),
+        ],
+        # inputs to the binded functions
+        operands=[mlir.ir_constant(l_max), normalized, mlir.ir_constant(n_sample), xyz],
+        # Layout specification:
+        operand_layouts=default_layouts((), (), (), x_shape),
+        result_layouts=default_layouts(shape, dshape),
+    )
+
+
+# register the custom xla lowering
+mlir.register_lowering(
+    _sph_fwd_p,
+    sph_lowering_cpu,
+    platform="cpu",
+)
+
+_sph_fwd_p.def_abstract_eval(sph_abstract_eval)
+
+
+### define how to compute the gradients backward
+# to be able to use dsph in sph_vjp without recomputing it is returned here
+# the drawback is that only backward diff. is possible with the current code
+# SUGGESTION change the bindings/Cpp API to be able to retrive the gradients
+# at a later stage
+def sph_jvp(l_max, normalized, xyz):
+    ls = jnp.arange(l_max + 1)
+    sph, dsph = sph_fwd(ls, normalized, xyz)
+    return sph, dsph
+
+
+def sph_vjp(l_max, normalized, args, tangents):
+    dsph = args[0]
+    out = jnp.einsum("ndl,nl->nd", dsph, tangents)
+    return (out,)
+
+
+@partial(jax.custom_vjp, nondiff_argnums=(0, 1))
+def spherical_harmonics(l_max, normalized, xyz):
+    """Computes the Spherical harmonics and their derivatives within
+    the JAX framework. see `sphericart.SphericalHarmonics` for more details.
+
+    Parameters
+    ----------
+    l_max : int
+        maximum order of the spherical harmonics (included)
+    normalized : bool
+        if the imput vectors have already been normalized
+    xyz : jax array [n_sample, 3]
+        set of n_sample vectors in 3D
+
+    Returns
+    -------
+    jax array [n_sample, (l_max+1)**2]
+        Spherical harmonics expension of `xyz`
+    """
+    ls = jnp.arange(l_max + 1)
+    output, _ = sph_fwd(ls, normalized, xyz)
+    return output
```

Comment:
It seems to me that we are evaluating both sph and dsph during the call to the spherical harmonics. That is however a large overhead if the user does not want to take derivatives. Would it be possible to calculate the derivatives only when people ask for computations that need the gradients?

Reply:
There is little point of using JAX if one does not need gradients. The numpy implementation would be a better fit at that point.

Reply:
I agree that people use jax for gradients, but they don't have to be gradients of the spherical harmonics. For example, when you train a NN on energies, you need the gradients with respect to the weights, but not with respect to the spherical harmonics. Think also about higher derivatives: we won't be able to assume that users want the second derivatives at some point in their workflow... the only choice would be to calculate them on the fly